parallel-exec/lib/lib.go

```diff
@@ -0,0 +1,418 @@
+		return "started"
+	case EventTypeCmdStarted:
+		return "cmd_started"
+	case EventTypeCmdFinished:
+		return "cmd_finished"
+	case EventTypeFinished:
+		return "finished"
+	default:
+		return strconv.Itoa(int(e))
+	}
+}
+
+// ParseEventType parses an EventType from it's string representation.
+func ParseEventType(eventTypeString string) (EventType, error) {
+	switch eventTypeString {
+	case "started":
+		return EventTypeStarted, nil
+	case "cmd_started":
+		return EventTypeCmdStarted, nil
+	case "cmd_finished":
+		return EventTypeCmdFinished, nil
+	case "finished":
+		return EventTypeFinished, nil
+	default:
+		return 0, fmt.Errorf("unknown EventType string: %s", eventTypeString)
+	}
+}
+
+// Event is an event that happens during the runner's run call.
+type Event interface {
+	Type() EventType
+	Time() time.Time
+	Fields() map[string]string
+}
+
+// EventHandler handles events.
+type EventHandler func(Event)
+
+// Timer returns the current time.
+type Timer func() time.Time
+
+// RunnerOption is an option for a new Runner.
+type RunnerOption func(*runnerOptions)
+
+// WithFastFail returns a RunnerOption that will return error fun
+// Run as soon as one of the commands fails.
+func WithFastFail() RunnerOption {
+	return func(runnerOptions *runnerOptions) {
+		runnerOptions.FastFail = true
+	}
+}
+
+// WithMaxConcurrentCmds returns a RunnerOption that will make the
+// Runner only run maxConcurrentCmds at once, or unlimited if 0.
+func WithMaxConcurrentCmds(maxConcurrentCmds int) RunnerOption {
+	return func(runnerOptions *runnerOptions) {
+		runnerOptions.MaxConcurrentCmds = maxConcurrentCmds
+	}
+}
+
+// WithEventHandler returns a RunnerOption that will use the
+// given EventHandler.
+func WithEventHandler(eventHandler EventHandler) RunnerOption {
+	return func(runnerOptions *runnerOptions) {
+		runnerOptions.EventHandler = eventHandler
+	}
+}
+
+// WithTimer returns a RunnerOption that will make the Runner
+// use the given Timer.
+func WithTimer(timer Timer) RunnerOption {
+	return func(runnerOptions *runnerOptions) {
+		runnerOptions.Timer = timer
+	}
+}
+
+// Runner runs the commands.
+type Runner interface {
+	// Run the commands.
+	//
+	// Return error if there was an initialization error, or any of
+	// the running commands returned with a non-zero exit code.
+	Run(cmds []*exec.Cmd) error
+}
+
+// NewRunner returns a new Runner.
+func NewRunner(options ...RunnerOption) Runner {
+	return newRunner(options...)
+}
+
+type event struct {
+	E EventType
+	T time.Time
+	F map[string]string
+}
+
+func newEvent(e EventType, t time.Time, f map[string]string) *event {
+	return &event{e, t, f}
+}
+
+func newStartedEvent(t time.Time) *event {
+	return newEvent(EventTypeStarted, t, nil)
+}
+
+func newCmdStartedEvent(t time.Time, cmd *exec.Cmd) *event {
+	return newEvent(EventTypeCmdStarted, t, map[string]string{
+		"cmd": cmdString(cmd),
+	})
+}
+
+func newCmdFinishedEvent(t time.Time, cmd *exec.Cmd, startTime time.Time, err error) *event {
```

Comment:
It would be, but the problem is I need both the time of the event and the duration.